File: app/src/main/java/com/streamarc/tv/transfer/TransferService.kt
package com.streamarc.tv.transfer

import android.app.AlarmManager
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.pm.ServiceInfo
import android.os.Build
import android.os.IBinder
import android.os.PowerManager
import androidx.core.app.NotificationCompat
import androidx.core.app.ServiceCompat
import androidx.core.content.ContextCompat
import com.streamarc.tv.R
import com.streamarc.tv.data.XtreamApi
import com.streamarc.tv.ui.MainActivity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit
import okhttp3.Request
import java.util.concurrent.TimeUnit

/**
 * Foreground service that performs downloads and live recordings: streams the
 * URL into the chosen folder, reports progress to [TransferStore], stops a
 * recording when its end time passes, and schedules future recordings with
 * AlarmManager.
 */
class TransferService : Service() {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val running = HashMap<String, Job>()
    private val downloadSlots = Semaphore(2)
    private lateinit var store: TransferStore
    private var wakeLock: PowerManager.WakeLock? = null

    override fun onBind(intent: Intent?): IBinder? = null

    override fun onCreate() {
        super.onCreate()
        store = TransferStore.get(this)
        ensureChannel()
        ServiceCompat.startForeground(
            this, NOTIF_ID, buildNotification(getString(R.string.transfers_working), null),
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) ServiceInfo.FOREGROUND_SERVICE_TYPE_DATA_SYNC else 0
        )
        wakeLock = (getSystemService(Context.POWER_SERVICE) as PowerManager)
            .newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "StreamArcTV:transfers").also { it.acquire(12 * 60 * 60 * 1000L) }
        scope.launch { progressLoop() }
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        when (intent?.action) {
            ACTION_CANCEL -> intent.getStringExtra(EXTRA_ID)?.let { cancel(it) }
            else -> kick()
        }
        return START_STICKY
    }

    /** Starts anything that is due: queued downloads and recordings whose time has come. */
    private fun kick() {
        val now = System.currentTimeMillis()
        for (job in store.all()) {
            if (running.containsKey(job.id)) continue
            when {
                job.state == TransferState.QUEUED -> start(job)
                job.state == TransferState.SCHEDULED && job.startAt <= now + 15_000 -> start(job)
                job.state == TransferState.SCHEDULED -> scheduleAlarm(this, job)
                job.state == TransferState.RUNNING -> start(job)   // the process died mid-transfer: restart it
            }
        }
        stopIfIdle()
    }

    private fun start(job: TransferJob) {
        running[job.id] = scope.launch {
            if (job.type == TransferType.DOWNLOAD) downloadSlots.withPermit { run(job) } else run(job)
            running.remove(job.id)
            stopIfIdle()
        }
    }

    private fun cancel(id: String) {
        running.remove(id)?.cancel()
        store.get(id)?.let { j ->
            Folders.delete(this, j.fileUri)
            store.update(id) { it.state = TransferState.CANCELLED }
        }
        cancelAlarm(this, id)
        stopIfIdle()
    }

    private suspend fun run(job: TransferJob) {
        val id = job.id
        // A recording scheduled for later waits here (the alarm normally wakes us anyway).
        if (job.type == TransferType.RECORDING) {
            val wait = job.startAt - System.currentTimeMillis()
            if (wait > 0) delay(wait)
        }
        store.update(id) { it.state = TransferState.RUNNING; it.error = null }
        val mime = if (job.fileName.endsWith(".ts", true)) "video/mp2t" else "video/mp4"
        var target: Folders.Target? = null
        try {
            target = Folders.create(this, job.folder, job.type, job.fileName, mime)
            store.update(id) { it.fileUri = target.uri.toString() }
            val client = XtreamApi.client.newBuilder()
                .readTimeout(60, TimeUnit.SECONDS)
                .callTimeout(0, TimeUnit.MILLISECONDS)
                .build()
            val req = Request.Builder().url(job.url).header("User-Agent", XtreamApi.USER_AGENT).build()
            client.newCall(req).execute().use { resp ->
                if (!resp.isSuccessful) throw IllegalStateException("Server returned HTTP ${resp.code}")
                val body = resp.body ?: throw IllegalStateException("Empty response")
                val total = if (job.type == TransferType.DOWNLOAD) body.contentLength() else -1L
                store.update(id) { it.total = total }
                val input = body.byteStream()
                val buf = ByteArray(256 * 1024)
                var done = 0L
                var lastFlush = System.currentTimeMillis()
                while (scope.isActive && running.containsKey(id)) {
                    if (job.type == TransferType.RECORDING && System.currentTimeMillis() >= job.endAt) break
                    val n = input.read(buf)
                    if (n < 0) break
                    target.stream.write(buf, 0, n)
                    done += n
                    val now = System.currentTimeMillis()
                    if (now - lastFlush > 1000) {
                        lastFlush = now
                        val d = done
                        store.update(id) { it.bytes = d }
                    }
                }
                target.stream.flush()
                val d = done
                store.update(id) { it.bytes = d }
            }
            if (!running.containsKey(id)) return   // cancelled
            store.update(id) { it.state = TransferState.DONE }
            notifyDone(job, true, null)
        } catch (e: Exception) {
            if (!running.containsKey(id)) return
            store.update(id) { it.state = TransferState.FAILED; it.error = e.message }
            notifyDone(job, false, e.message)
        } finally {
            try { target?.stream?.close() } catch (_: Exception) {}
        }
    }

    private suspend fun progressLoop() {
        while (scope.isActive) {
            val active = store.all().filter { it.state == TransferState.RUNNING }
            val text = when (active.size) {
                0 -> getString(R.string.transfers_idle)
                1 -> active[0].let { j ->
                    val pct = if (j.total > 0) " ${(j.bytes * 100 / j.total)}%" else ""
                    "${if (j.type == TransferType.RECORDING) getString(R.string.recording) else getString(R.string.downloading)}: ${j.title}$pct"
                }
                else -> getString(R.string.transfers_count_fmt, active.size)
            }
            val nm = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            nm.notify(NOTIF_ID, buildNotification(text, active.firstOrNull()))
            delay(2000)
        }
    }

    private fun stopIfIdle() {
        scope.launch {
            delay(1500)
            val pending = store.all().any { it.state == TransferState.QUEUED || it.state == TransferState.RUNNING } || running.isNotEmpty()
            if (!pending) stopSelf()
        }
    }

    override fun onDestroy() {
        scope.cancel()
        try { wakeLock?.let { if (it.isHeld) it.release() } } catch (_: Exception) {}
        super.onDestroy()
    }

    private fun ensureChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val nm = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            nm.createNotificationChannel(NotificationChannel(CHANNEL, getString(R.string.transfers_channel), NotificationManager.IMPORTANCE_LOW))
        }
    }

    private fun buildNotification(text: String, job: TransferJob?): Notification {
        val open = PendingIntent.getActivity(
            this, 0, Intent(this, MainActivity::class.java),
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
        val b = NotificationCompat.Builder(this, CHANNEL)
            .setSmallIcon(R.drawable.ic_update)
            .setContentTitle(getString(R.string.app_name))
            .setContentText(text)
            .setOngoing(true)
            .setOnlyAlertOnce(true)
            .setContentIntent(open)
        if (job != null && job.total > 0) b.setProgress(1000, (job.bytes * 1000 / job.total).toInt(), false)
        else if (job != null) b.setProgress(0, 0, true)
        return b.build()
    }

    private fun notifyDone(job: TransferJob, ok: Boolean, error: String?) {
        val nm = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val text = if (ok) getString(R.string.transfer_done_fmt, job.title) else getString(R.string.transfer_failed_fmt, job.title, error ?: "")
        val n = NotificationCompat.Builder(this, CHANNEL)
            .setSmallIcon(R.drawable.ic_update)
            .setContentTitle(getString(R.string.app_name))
            .setContentText(text)
            .setAutoCancel(true)
            .build()
        nm.notify(job.id.hashCode(), n)
    }

    companion object {
        const val CHANNEL = "transfers"
        const val NOTIF_ID = 4101
        const val ACTION_CANCEL = "com.streamarc.tv.transfer.CANCEL"
        const val EXTRA_ID = "id"

        /** Adds a job and makes sure the service is running (or the alarm is set). */
        fun enqueue(context: Context, job: TransferJob) {
            TransferStore.get(context).put(job)
            if (job.state == TransferState.SCHEDULED && job.startAt > System.currentTimeMillis() + 60_000) {
                scheduleAlarm(context, job)
            } else {
                startService(context)
            }
        }

        fun startService(context: Context) {
            ContextCompat.startForegroundService(context, Intent(context, TransferService::class.java))
        }

        fun cancel(context: Context, id: String) {
            val job = TransferStore.get(context).get(id)
            if (job != null && job.state == TransferState.SCHEDULED) {
                cancelAlarm(context, id)
                Folders.delete(context, job.fileUri)
                TransferStore.get(context).update(id) { it.state = TransferState.CANCELLED }
                return
            }
            val i = Intent(context, TransferService::class.java).setAction(ACTION_CANCEL).putExtra(EXTRA_ID, id)
            ContextCompat.startForegroundService(context, i)
        }

        private fun alarmIntent(context: Context, id: String): PendingIntent =
            PendingIntent.getBroadcast(
                context, id.hashCode(),
                Intent(context, TransferAlarmReceiver::class.java).setAction("com.streamarc.tv.transfer.ALARM").putExtra(EXTRA_ID, id),
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
            )

        fun scheduleAlarm(context: Context, job: TransferJob) {
            val am = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
            val pi = alarmIntent(context, job.id)
            val at = job.startAt - 5_000
            val exactOk = Build.VERSION.SDK_INT < Build.VERSION_CODES.S || am.canScheduleExactAlarms()
            try {
                if (exactOk) {
                    am.setAlarmClock(AlarmManager.AlarmClockInfo(at, pi), pi)
                } else {
                    am.setWindow(AlarmManager.RTC_WAKEUP, at, 60_000, pi)
                }
            } catch (_: SecurityException) {
                am.setWindow(AlarmManager.RTC_WAKEUP, at, 60_000, pi)
            }
        }

        fun cancelAlarm(context: Context, id: String) {
            val am = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
            am.cancel(alarmIntent(context, id))
        }

        /** Re-arms alarms for scheduled recordings (after reboot or app update). */
        fun rescheduleAll(context: Context) {
            val now = System.currentTimeMillis()
            for (job in TransferStore.get(context).all()) {
                if (job.state == TransferState.SCHEDULED) {
                    if (job.endAt <= now) TransferStore.get(context).update(job.id) { it.state = TransferState.FAILED; it.error = "Missed" }
                    else if (job.startAt <= now) startService(context)
                    else scheduleAlarm(context, job)
                }
            }
        }
    }
}

/** Woken by AlarmManager for a scheduled recording, or by boot to re-arm alarms. */
class TransferAlarmReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        when (intent.action) {
            Intent.ACTION_BOOT_COMPLETED, Intent.ACTION_MY_PACKAGE_REPLACED -> TransferService.rescheduleAll(context)
            else -> TransferService.startService(context)
        }
    }
}

File: windows/StreamArcTV/Transfer/TransferService.cs
using System.Collections.Concurrent;
using System.Diagnostics;
using System.IO;
using System.Net.Http;
using System.Net.NetworkInformation;
using StreamArcTV.Data;
using StreamArcTV.Util;

namespace StreamArcTV.Transfer;

/// <summary>
/// Background engine that performs downloads and live recordings: streams the URL into the
/// chosen folder, reports progress to <see cref="TransferStore"/>, stops a recording when its end
/// time passes, and schedules future recordings (in-process while the app runs, and as a Windows
/// scheduled task that starts the app in the background when it is closed).
/// </summary>
public static class TransferService
{
    private static readonly ConcurrentDictionary<string, CancellationTokenSource> Running = new();
    // Xtream panels normally allow one connection per account, so downloads run strictly one
    // after another; a whole season is a queue, not a burst.
    private static readonly SemaphoreSlim DownloadSlots = new(1);
    private static readonly TransferStore Store = TransferStore.Get();
    private static System.Threading.Timer? _scheduler;

    private const int MAX_ATTEMPTS = 6;
    private static readonly long[] RETRY_DELAYS_MS = { 5_000, 15_000, 30_000, 60_000, 120_000 };

    /// True while the player is open; downloads hold off so the stream gets the connection.
    public static volatile bool PlaybackActive;

    /// Live progress by job id; only present while the transfer is actually moving bytes.
    public static readonly ConcurrentDictionary<string, Progress> Live = new();

    /// Up-to-the-chunk progress of a running transfer, read directly by the lists.
    public class Progress
    {
        public long Bytes;
        public long Total;
        public double BytesPerSec;
        public Progress(long bytes, long total) { Bytes = bytes; Total = total; }
    }

    /// Fired (on a worker thread) when a transfer finishes or fails, for the tray notification.
    public static event Action<TransferJob, bool, string?>? Finished;

    public static bool AnyActive => Running.Count > 0 || Store.All().Any(j => j.State is TransferState.QUEUED or TransferState.RUNNING);

    /// Adds a job and makes sure it is started (or scheduled).
    public static void Enqueue(TransferJob job)
    {
        Store.Put(job);
        if (job.State == TransferState.SCHEDULED && job.StartAt > Format.NowMs + 60_000) ScheduleTask(job);
        Kick();
    }

    /// Starts anything that is due: queued downloads and recordings whose time has come.
    public static void Kick()
    {
        _scheduler ??= new System.Threading.Timer(_ => Kick(), null, 15_000, 15_000);
        var now = Format.NowMs;
        foreach (var job in Store.All())
        {
            if (Running.ContainsKey(job.Id)) continue;
            switch (job.State)
            {
                case TransferState.QUEUED: Start(job); break;
                case TransferState.SCHEDULED when job.StartAt <= now + 15_000: Start(job); break;
                case TransferState.RUNNING: Start(job); break;   // the process died mid-transfer: restart it
            }
        }
    }

    /// Re-arms scheduled recordings on start-up: missed ones fail, due ones start, later ones are scheduled.
    /// Renames the files of downloads completed by earlier versions ("Show S1E2 Title.mp4") to the
    /// clear form ("Show - S01E02 - Title.mp4") and updates their entries. Runs once, in the background.
    public static void RenameCompleted()
    {
        if (Prefs.Instance.DownloadsRenamed) return;
        Task.Run(() =>
        {
            var renamed = 0;
            foreach (var job in Store.All())
            {
                if (job.Type != TransferType.DOWNLOAD || job.State != TransferState.DONE) continue;
                var title = Names.Upgrade(job.Title);
                if (title == null || title == job.Title) continue;
                var ext = Path.GetExtension(job.FileName);
                var fileName = Folders.SafeName(title) + ext;
                var newPath = job.FileUri;
                try
                {
                    if (job.FileUri != null && File.Exists(job.FileUri))
                    {
                        newPath = Path.Combine(Path.GetDirectoryName(job.FileUri)!, fileName);
                        if (!string.Equals(newPath, job.FileUri, StringComparison.OrdinalIgnoreCase))
                        {
                            if (File.Exists(newPath)) File.Delete(newPath);
                            File.Move(job.FileUri, newPath);
                        }
                    }
                }
                catch (Exception e) { AppLog.W("Transfer", $"rename '{job.FileName}' failed: {e.Message}"); continue; }
                Store.Update(job.Id, j => { j.Title = title; j.FileName = fileName; j.FileUri = newPath; });
                renamed++;
            }
            if (renamed > 0) AppLog.I("Transfer", $"renamed {renamed} completed download(s) to the clear form");
            Prefs.Instance.DownloadsRenamed = true;
        });
    }

    public static void RescheduleAll()
    {
        var now = Format.NowMs;
        foreach (var job in Store.All())
        {
            if (job.State != TransferState.SCHEDULED) continue;
            if (job.EndAt <= now) Store.Update(job.Id, j => { j.State = TransferState.FAILED; j.Error = "Missed"; });
            else if (job.StartAt > now + 60_000) ScheduleTask(job);
        }
        Kick();
    }

    private static void Start(TransferJob job)
    {
        var cts = new CancellationTokenSource();
        if (!Running.TryAdd(job.Id, cts)) return;
        _ = Task.Run(async () =>
        {
            try
            {
                if (job.Type == TransferType.DOWNLOAD)
                {
                    await DownloadSlots.WaitAsync(cts.Token);
                    try { await Run(job, cts.Token); } finally { DownloadSlots.Release(); }
                }
                else await Run(job, cts.Token);
            }
            catch (OperationCanceledException) { }
            catch (Exception e) { AppLog.E("Transfer", $"'{job.Title}' crashed", e); }
            finally { Running.TryRemove(job.Id, out _); }
        });
    }

    public static void Cancel(string id)
    {
        var job = Store.Get(id);
        if (Running.TryRemove(id, out var cts)) cts.Cancel();
        if (job != null)
        {
            Folders.Delete(job.FileUri);
            Store.Update(id, j => j.State = TransferState.CANCELLED);
        }
        UnscheduleTask(id);
    }

    private static async Task Run(TransferJob job, CancellationToken ct)
    {
        var id = job.Id;
        // A recording scheduled for later waits here (the scheduler normally wakes us anyway).
        if (job.Type == TransferType.RECORDING)
        {
            var wait = job.StartAt - Format.NowMs;
            if (wait > 0) await Task.Delay((int)Math.Min(wait, int.MaxValue), ct);
        }
        AppLog.I("Transfer", $"start {job.Type} '{job.Title}' {AppLog.SafeUrl(job.Url)}");
        Store.Update(id, j => { j.State = TransferState.RUNNING; j.Error = null; });
        Folders.Target? target = null;
        long done = 0;
        var attempt = 0;
        try
        {
            while (!ct.IsCancellationRequested && Running.ContainsKey(id))
            {
                // Playback has priority: most providers allow one stream per account, so a
                // download would stop the video from loading. Wait here until the player closes.
                if (job.Type == TransferType.DOWNLOAD && HoldReason() != null)
                {
                    AppLog.I("Transfer", $"'{job.Title}' held ({HoldReason()}) at {done} bytes");
                    Store.Update(id, j => j.Error = HoldReason() == "wifi" ? "Waiting for Wi-Fi (mobile data is off for downloads)" : "Paused while you watch – resumes when the player closes");
                    while (HoldReason() != null && !ct.IsCancellationRequested && Running.ContainsKey(id)) await Task.Delay(1000, ct);
                    Store.Update(id, j => j.Error = null);
                    continue;
                }
                try
                {
                    var req = new HttpRequestMessage(HttpMethod.Get, job.Url);
                    req.Headers.UserAgent.ParseAdd(XtreamApi.USER_AGENT);
                    req.Headers.ConnectionClose = true;
                    var resuming = job.Type == TransferType.DOWNLOAD && done > 0 && target != null;
                    if (resuming) req.Headers.Range = new System.Net.Http.Headers.RangeHeaderValue(done, null);
                    using var pauseCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
                    using var resp = await HttpRedirects.SendAsync(Client, req, HttpCompletionOption.ResponseHeadersRead, pauseCts.Token);
                    if (!resp.IsSuccessStatusCode)
                    {
                        // Capture exactly what the provider answered so the reason is in the log and the list.
                        string detail = "";
                        try
                        {
                            var body = await resp.Content.ReadAsStringAsync(ct);
                            detail = System.Text.RegularExpressions.Regex.Replace(body, "<[^>]+>", " ").Trim();
                            if (detail.Length > 160) detail = detail[..160] + "…";
                        }
                        catch { }
                        AppLog.W("Transfer", $"'{job.Title}' HTTP {(int)resp.StatusCode} {resp.ReasonPhrase} for {AppLog.SafeUrl(job.Url)} · {detail}");
                        throw new HttpException((int)resp.StatusCode, resp.ReasonPhrase, detail);
                    }
                    var code = (int)resp.StatusCode;
                    Folders.Target? outT = null;
                    if (resuming && code == 206)
                    {
                        // Carry on where the last attempt stopped.
                        try { target?.Dispose(); } catch { }
                        try { outT = Folders.Append(target!.Path); } catch { outT = null; }
                    }
                    if (outT == null)
                    {
                        try { target?.Dispose(); } catch { }
                        outT = Folders.Create(job.Folder, job.Type, job.FileName);
                        done = 0;
                        if (resuming && code == 206) throw new InvalidOperationException("Can't resume file");
                    }
                    target = outT;
                    var path = outT.Path;
                    var len = resp.Content.Headers.ContentLength ?? -1;
                    var total = job.Type != TransferType.DOWNLOAD ? -1L : (code == 206 && len >= 0 ? len + done : len);
                    Store.Update(id, j => { j.FileUri = path; j.Total = total; j.Error = null; });
                    await using var input = await resp.Content.ReadAsStreamAsync(pauseCts.Token);
                    var buf = new byte[256 * 1024];
                    var prog = new Progress(done, total);
                    Live[id] = prog;
                    var lastFlush = Format.NowMs;
                    var lastFlushBytes = done;
                    var paused = false;
                    while (!ct.IsCancellationRequested && Running.ContainsKey(id))
                    {
                        if (job.Type == TransferType.RECORDING && Format.NowMs >= job.EndAt) break;
                        if (job.Type == TransferType.DOWNLOAD && HoldReason() != null)
                        {
                            // Close the socket for real, so the provider frees this account's stream slot for the player.
                            paused = true;
                            pauseCts.Cancel();
                            break;
                        }
                        using var readCts = CancellationTokenSource.CreateLinkedTokenSource(pauseCts.Token);
                        readCts.CancelAfter(60_000);
                        int n;
                        try { n = await input.ReadAsync(buf, readCts.Token); }
                        catch (OperationCanceledException) when (!pauseCts.IsCancellationRequested) { throw new TimeoutException("The provider stopped sending data (timed out)."); }
                        if (n <= 0) break;
                        outT.Stream.Write(buf, 0, n);
                        done += n;
                        prog.Bytes = done;
                        var now = Format.NowMs;
                        if (now - lastFlush > 1000)
                        {
                            prog.BytesPerSec = (done - lastFlushBytes) * 1000.0 / (now - lastFlush);
                            lastFlush = now;
                            lastFlushBytes = done;
                            var d = done;
                            Store.Update(id, j => j.Bytes = d);
                        }
                    }
                    outT.Stream.Flush();
                    var dd = done;
                    Store.Update(id, j => j.Bytes = dd);
                    if (paused) continue;   // back to the top of the loop, which waits and then resumes with a Range request
                    if (job.Type == TransferType.DOWNLOAD && total > 0 && done < total && Running.ContainsKey(id) && !ct.IsCancellationRequested)
                        throw new InvalidOperationException("Connection closed early");
                    AppLog.I("Transfer", $"'{job.Title}' done, {done} bytes");
                    if (!Running.ContainsKey(id) || ct.IsCancellationRequested) return;   // cancelled
                    Store.Update(id, j => { j.State = TransferState.DONE; j.Error = null; });
                    NotifyDone(job, true, null);
                    return;
                }
                catch (OperationCanceledException) when (ct.IsCancellationRequested) { return; }
                catch (Exception e)
                {
                    if (HoldReason() != null && job.Type == TransferType.DOWNLOAD) continue;   // the cancel above surfaces as an exception
                    if (!Running.ContainsKey(id) || ct.IsCancellationRequested) return;
                    attempt++;
                    if (attempt >= MAX_ATTEMPTS) throw;
                    var wait = RETRY_DELAYS_MS[Math.Min(attempt - 1, RETRY_DELAYS_MS.Length - 1)];
                    var why = Describe(e);
                    AppLog.W("Transfer", $"'{job.Title}' attempt {attempt} failed at {done} bytes: {why} ({e.GetType().Name}: {e.Message}); retry in {wait / 1000}s");
                    Store.Update(id, j => j.Error = $"Retrying ({attempt} of {MAX_ATTEMPTS}) in {wait / 1000} s – {why}");
                    await Task.Delay((int)wait, ct);
                }
            }
        }
        catch (OperationCanceledException) { }
        catch (Exception e)
        {
            if (!Running.ContainsKey(id)) return;
            var why = Describe(e);
            AppLog.E("Transfer", $"'{job.Title}' gave up: {why}", e);
            if (job.Type == TransferType.DOWNLOAD)
            {
                // A download that gave up leaves nothing behind: drop the partial file and the
                // list entry; the notification carries the reason.
                try { target?.Dispose(); } catch { }
                Folders.Delete(target?.Path);
                Store.Remove(id);
            }
            else
            {
                Store.Update(id, j => { j.State = TransferState.FAILED; j.Error = why; });
            }
            NotifyDone(job, false, why);
        }
        finally
        {
            Live.TryRemove(id, out _);
            try { target?.Dispose(); } catch { }
        }
    }

    /// Streaming client with no overall timeout (reads are watched individually above).
    private static readonly HttpClient Client = new(new SocketsHttpHandler
    {
        AllowAutoRedirect = false,   // followed by hand, see HttpRedirects
        ConnectTimeout = TimeSpan.FromSeconds(20),
        PooledConnectionIdleTimeout = TimeSpan.FromSeconds(2),
    })
    { Timeout = Timeout.InfiniteTimeSpan };

    /// Why a download must wait right now: "playback", "wifi", or null to go ahead.
    private static string? HoldReason()
    {
        if (PlaybackActive) return "playback";
        if (Prefs.Instance.DownloadsWifiOnly && OnMobileData()) return "wifi";
        return null;
    }

    /// True when the only way out is a mobile-broadband (WWAN) or dial-up adapter.
    public static bool OnMobileData()
    {
        try
        {
            var up = NetworkInterface.GetAllNetworkInterfaces()
                .Where(n => n.OperationalStatus == OperationalStatus.Up && n.NetworkInterfaceType != NetworkInterfaceType.Loopback && n.NetworkInterfaceType != NetworkInterfaceType.Tunnel)
                .Where(n => n.GetIPProperties().GatewayAddresses.Count > 0)
                .ToList();
            if (up.Count == 0) return false;
            return up.All(n => n.NetworkInterfaceType is NetworkInterfaceType.Wwanpp or NetworkInterfaceType.Wwanpp2 or NetworkInterfaceType.Ppp);
        }
        catch { return false; }
    }

    private class HttpException : Exception
    {
        public int Code { get; }
        public string Reason { get; }
        public string Detail { get; }
        public HttpException(int code, string? reason, string detail) : base($"HTTP {code} {reason}".Trim()) { Code = code; Reason = reason ?? ""; Detail = detail; }
        public string Label => $"HTTP {Code}{(Reason.Length > 0 ? " " + Reason : "")}{(Detail.Length > 0 ? ": " + Detail : "")}";
    }

    private static string Describe(Exception e) => e switch
    {
        HttpException { Code: 401 or 403 or 429 or 458 or 503 or 509 } h =>
            $"Provider refused the connection ({h.Label}), usually because too many streams are open on this account. It is retried automatically; pause other playback if it keeps failing.",
        HttpException { Code: 404 } h => $"The provider has no file for this title ({h.Label}).",
        HttpException h => $"Provider error ({h.Label}).",
        _ => string.IsNullOrWhiteSpace(e.Message) ? e.GetType().Name : e.Message
    };

    private static void NotifyDone(TransferJob job, bool ok, string? error)
    {
        try { Finished?.Invoke(job, ok, error); } catch { }
    }

    // ---- Windows scheduled tasks (recordings start even when the app is closed) ----

    private static string TaskName(string id) => $"StreamArcTV Recording {id}";

    /// Registers a scheduled task that starts the app in the background just before the recording.
    public static void ScheduleTask(TransferJob job)
    {
        try
        {
            var exe = Environment.ProcessPath ?? "";
            if (string.IsNullOrEmpty(exe)) return;
            var at = DateTimeOffset.FromUnixTimeMilliseconds(job.StartAt - 5_000).LocalDateTime;
            var script =
                $"$a = New-ScheduledTaskAction -Execute '{Ps(exe)}' -Argument '--tray'; " +
                $"$t = New-ScheduledTaskTrigger -Once -At ([datetime]'{at:yyyy-MM-ddTHH:mm:ss}'); " +
                $"$s = New-ScheduledTaskSettingsSet -AllowStartIfOnBatteries -DontStopIfGoingOnBatteries -StartWhenAvailable; " +
                $"Register-ScheduledTask -TaskName '{Ps(TaskName(job.Id))}' -Action $a -Trigger $t -Settings $s -Force | Out-Null";
            RunPowerShell(script);
        }
        catch (Exception e) { AppLog.W("Transfer", $"couldn't register the scheduled task: {e.Message}"); }
    }

    public static void UnscheduleTask(string id)
    {
        try { RunPowerShell($"Unregister-ScheduledTask -TaskName '{Ps(TaskName(id))}' -Confirm:$false -ErrorAction SilentlyContinue"); }
        catch { }
    }

    private static string Ps(string s) => s.Replace("'", "''");

    private static void RunPowerShell(string script)
    {
        var psi = new ProcessStartInfo("powershell.exe", $"-NoProfile -NonInteractive -WindowStyle Hidden -Command \"{script.Replace("\"", "\\\"")}\"")
        {
            CreateNoWindow = true, UseShellExecute = false, RedirectStandardError = true, RedirectStandardOutput = true
        };
        using var p = Process.Start(psi);
        if (p == null) return;
        var err = p.StandardError.ReadToEnd();
        p.WaitForExit(20_000);
        if (p.ExitCode != 0 && !string.IsNullOrWhiteSpace(err)) AppLog.W("Transfer", "powershell: " + err.Trim());
    }
}
